crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1654,48 +1654,56 @@ impl<'db> TypeInferenceBuilder<'db> {
     fn infer_fstring_expression(&mut self, fstring: &ast::ExprFString) -> Type<'db> {
         let ast::ExprFString { range: _, value } = fstring;
 
+        let mut has_expression = false;
+        let mut literals = Vec::new();
+        // Build an arena to allocate expression string representation
+        let expr_arena = typed_arena::Arena::new();
         for part in value {
             match part {
-                ast::FStringPart::Literal(_) => {
-                    // TODO string literal type
+                ast::FStringPart::Literal(literal) => {
+                    literals.push(&literal.value);
                 }
                 ast::FStringPart::FString(fstring) => {
-                    let ast::FString {
-                        range: _,
-                        elements,
-                        flags: _,
-                    } = fstring;
-                    for element in elements {
-                        self.infer_fstring_element(element);
+                    for element in &fstring.elements {
+                        match element {
+                            ast::FStringElement::Expression(expression) => {
+                                let ty = self.infer_expression(&expression.expression);
+                                match ty {
+                                    Type::BooleanLiteral(_) | Type::IntLiteral(_) => {
+                                        let repr = format!("{}", ty.representation(self.db));
```

Comment:
Using the display representation to get the value of the type is dangerous. We could accidentally change the inferred literal value by changing how those types are displayed.

Reply:
That's fair - would you re-code the representation here then or just not provide this feature for bool/int?

Reply:
That's fair - would you re-code the representation here then or just not provide this feature for bool/int?

Reply:
I don't mind the feature, although it probably goes a bit out of scope for this PR :) 

I would re-code it or even add a dedicated method on `Type` because it is important that the value returned from this method exactly matches the runtime semantics.

Reply:
Just to be clear, what do you mean by "the runtime semantics"? Is there a source I can refer to to know what's the "correct" display for a given int/bool?

Reply:
Yes, we could have a `Type::repr()` method that returns `Option<String>`, that attempts to exactly emulate the result of callling `repr()` on an instance of the type at runtime in Python (or returns `None` if we can't statically determine what the repr would be). This might be different from the `representation` method in some cases, as the `representation` method is just an internal implementation detail of how we implement `Display` for various types. Using `representation` works fine for now, but could easily break in the future.

Reply:
Although for unions, for example, we might have multiple possible `repr()`s of a value, so maybe `Option<Vec<String>>` would be a better return type for `repr`. E.g. calling repr on `Type::Union([StringLiteral("foo"), StringLiteral("bar")])` might return `Some([String("foo"), String("bar")])` -- we know the repr of an object of that type will be _one_ of those two strings

Reply:
I'm not sure if I would go as far as returning a `Vec`. It's unclear how and where we would use that. But right, `repr` is the behavior we need :)

Reply:
Consider this code:

```py
from typing import Literal

def foo(x: Literal["foo", "bar"]):
    y = f"{x}{x}"
```

The repr of `x` will be either `"foo"` or `"bar"`. If `Type::repr` returned a vec of possible reprs, we could reasonably infer that the value of `y` will be either `"foofoo"` or `"barbar"`, meaning the type of y will be `Literal["foofoo", "barbar"]`.

Reply:
(@JelleZijlstra who is sitting next to me at the core dev sprint informs me that his type checker, pyanalyze, already has this feature 😄)

Reply:
We could but I think it's going a bit too far. You very quickly end up with quadratic complexity and I'm not sure how useful the precise type is when refining. 

I'm looking for actual code patterns that this enables.

Reply:
Does he find it to be a useful feature?

Reply:
I'm fine returning `None` for unions for now. You're right that it is quadratic in behaviour, and Jelle doesn't think it's a crucial feature 😄

Reply:
I think we do want `Type::repr`, and we should use it here. I think it's fine to either entirely leave that as a TODO in this PR, or it's also fine to add a simple `Type::repr` in this PR that handles a few simple types and leaves the rest as TODO. (Main advantage of the latter is that you already have tests in this PR for f-strings including ints and booleans, and it will be simpler to keep them than have to take them out and re-add them later.)